src/Jackalope/Version/VersionHandler.php

```diff
@@ -0,0 +1,153 @@
+        $additionalOperations[] = new AddNodeOperation($rootVersion->getPath(), $rootVersion);
+
+        // TODO add frozen node to root version
+
+        $node->setProperty('jcr:versionHistory', $versionHistory);
+        $node->setProperty('jcr:baseVersion', $rootVersion); // TODO set correct base version
+        $node->setProperty('jcr:predecessors', array($rootVersion));
+
+        return $additionalOperations;
+    }
+
+    /**
+     * Performs a checkin for the node on the given path
+     * @param $path
+     * @return string
+     */
+    public function checkinItem($path)
+    {
+        $node = $this->objectManager->getNodeByPath($path);
+
+        if (!$node->isNodeType(static::MIX_SIMPLE_VERSIONABLE)) {
+            throw new UnsupportedRepositoryOperationException(
+                'Node has to implement at least "mix:versionable" to use verisoning operations'
+            );
+        }
+
+        if ($node->isModified()) {
+            throw new InvalidItemStateException(sprintf(
+                'Node "%s" contains unsaved changes',
+                $path
+            ));
+        }
+
+        if (!$node->isCheckedOut()) {
+            return $path;
+        }
+
+        if ($node->hasProperty('jcr:mergeFailed')) {
+            throw new VersionException(sprintf('Node "%s" contains unresolved merge conflicts', $path));
+        }
+
+        // TODO set subgraph to read only
+
+        $versionHistoryNode = $node->getPropertyValue('jcr:versionHistory');
+
+        // FIXME add some kind of sharding
+        $versionNode = $versionHistoryNode->addNode(UUIDHelper::generateUUID(), 'nt:version');
+        $versionNode->setProperty('jcr:uuid', UUIDHelper::generateUUID());
+        $versionNode->setProperty('jcr:created', new \DateTime());
+
+        // TODO create frozen node
+
+        $versionNode->setProperty('jcr:predecessors', $node->getProperty('jcr:predecessors')->getString());
+        $node->setProperty('jcr:predecessors', array(), PropertyType::REFERENCE, false);
+        foreach ($versionNode->getProperty('jcr:predecessors') as $predecessorUuid) {
+            $predecessor = $this->objectManager->getNodeByIdentifier($predecessorUuid);
+            $predecessor->setProperty('jcr:successors', array_merge($predecessor->getPropertyValueWithDefault('jcr:successors', array()), array($versionNode)));
+        }
+
+        // TODO change base version
+
+        $node->setProperty('jcr:isCheckedOut', false, PropertyType::BOOLEAN, false);
+
+        $this->session->save();
+
+        return $versionNode->getPath();
+    }
+
+    public function checkoutItem($path)
+    {
+        $node = $this->objectManager->getNodeByPath($path);
+
+        if ($node->isCheckedOut()) {
+            return;
+        }
+
+        if (!$node->isNodeType(static::MIX_SIMPLE_VERSIONABLE)) {
+            throw new UnsupportedRepositoryOperationException(
+                'Node has to implement at least "mix:versionable" to use verisoning operations'
```

Comment:
type: "verisoning". and i think you should say "at least mix:simpleVersionable" no?